Add free tier banners

zarsky-broad opened the pull request:
Requires matching PR in firecloud-develop

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: If you changed a URL that is used elsewhere (e.g. in an email), comment about where it is used and ensure the dependent code is updated.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [ ] **Submitter**: If you're adding new automated UI tests, review the test plan with QA
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Submitter**: Verify all tests go green, including CI tests and automated UI tests.
- [ ] **Submitter**: Squash commits and merge to develop. If adding test code, merge application code and test code at the same time.
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

zarsky-broad committed:
make user profile a shared atom

zarsky-broad committed:
add trial alerts

zarsky-broad committed:
centralize profile reload method, allow hiding banner

zarsky-broad committed:
allow warning color, show loading indicator

zarsky-broad committed:
Merge branch 'develop' into iz-GAWB-2810-free-tier-banners

zarsky-broad committed:
Merge branch 'develop' into iz-GAWB-2810-free-tier-banners

zarsky-broad committed:
decomponentized spinner

zarsky-broad committed:
fix for showing spinner

zarsky-broad committed:
spinner accepts attributes or text in 1-arity

zarsky-broad committed:
better indentation, add test-ids

zarsky-broad committed:
automation - keep custom headers with json requests

zarsky-broad committed:
automation - add get user profile method

zarsky-broad committed:
automation - add thurloe set key method

zarsky-broad committed:
automation - add FreeTrialSpec

still needs to test activating trial

zarsky-broad committed:
automation - test that starting trial works

zarsky-broad committed:
get dockerhost address better

zarsky-broad committed:
update akka and akka-http

zarsky-broad commented:
jenkins retest

zarsky-broad commented:
jenkins retest

zarsky-broad commented:
jenkins retest

zarsky-broad commented:
jenkins retest

zarsky-broad commented:
Well, it passed when running manually against a fiab. I think jenkins isn't pulling in the fc-develop branch correctly.

aednichols commented on automation/docker/run-tests.sh:
```diff
@@ -21,7 +21,7 @@ ENV=dev
 TEST_ENTRYPOINT="testOnly -- -l ProdTest"
 TEST_CONTAINER="automation-$(head /dev/urandom | env LC_CTYPE=C tr -dc a-z0-9 | head -c 8)"
-DOCKERHOST_ADDRESS=$(docker network inspect docker_default | grep Gateway | awk -F '"' '{ print $4 }')
+DOCKERHOST_ADDRESS=$(docker network inspect docker_default -f='{{(index .IPAM.Config 0).Gateway}}')
```
Could you explain what this does in a comment? I think I remember from discussing it, but that won't benefit anyone else

aednichols commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/api/FireCloudClient.scala:
```diff
@@ -78,7 +78,7 @@ trait FireCloudClient {
 
   private def requestWithJsonContent(method: HttpMethod, uri: String, content: Any, httpHeaders: List[HttpHeader] = List())(implicit token: AuthToken): String = {
-    val req = HttpRequest(method, uri, List(makeAuthHeader(token)), HttpEntity(ContentTypes.`application/json`, mapper.writeValueAsString(content)))
+    val req = HttpRequest(method, uri, List(makeAuthHeader(token)) ++ httpHeaders, HttpEntity(ContentTypes.`application/json`, mapper.writeValueAsString(content)))
```
Nice find. It seems there exists a `"-Ywarn-unused:params"`, but in a brief exploration I couldn't readily figure out how to have it take effect

aednichols commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/FreeTrialSpec.scala:
```diff
@@ -0,0 +1,76 @@
+
+      withSignIn(testUser) { _ =>
+        await ready new WorkspaceListPage()
+        val bannerTitleEl = Label(TestId("trial-banner-title"))
```
I assume this is saying `bannerTitleEl[igible]` but I don't think that's more helpful than `bannerTitle`

aednichols commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/api/Thurloe.scala:
```diff
@@ -40,6 +40,12 @@ object Thurloe extends FireCloudClient with LazyLogging {
 
+    def set(subjectId: String, key: String, value: String)(implicit token: AuthToken): Unit = {
+      logger.info(s"Setting $key as $value for $subjectId")
+      postRequest(url + s"api/thurloe", Map("userId" -> subjectId, "keyValuePairs" -> List(Map("key" -> key, "value" -> value))),
```
Nit: if you're going to go to multiple lines, I would break before the `Map` also because it's far-and-away the longest

davidangb replied:
also do you really want to keep the logger.info in here?

zarsky-broad replied:
@davidangb I could absolutely remove it, we seem to be logging things like this to the console. If we want to remove some, it'll significantly decrease the noise when viewing test results.

aednichols commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/FreeTrialSpec.scala:
```diff
@@ -0,0 +1,76 @@
+  val adminUser: Credentials = UserPool.chooseAdmin
+  implicit val authToken: AuthToken = adminUser.makeAuthToken()
+
+  val testUser: Credentials = Config.Users.testUser  // TODO: pull from user pool and fetch correct subject ID
```
This TODO appears to be left over from `RegistrationSpec`. Is it still valid?

aednichols commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/FreeTrialSpec.scala:
```diff
@@ -0,0 +1,76 @@
+
+  val testUser: Credentials = Config.Users.testUser  // TODO: pull from user pool and fetch correct subject ID
+  val userAuthToken: AuthToken = testUser.makeAuthToken()
+  var subjectId : String = Orchestration.profile.getUser()(userAuthToken)("userId").toString // TODO: Use orch endpoint to change trial status
```
Reference GAWB-2808

aednichols reviewed (commented):
Left some notes on the Scala tests. ~I can't get the UI to load up for me (screenshot), so haven't dug in to the CLJS.~ Hmm, I think it's configs, but it's not manifesting itself in the usual way

![screen shot 2017-12-04 at 12 47 09 pm](https://user-images.githubusercontent.com/1087943/33567671-fa5c2a02-d8f1-11e7-82d0-2120c415e697.png)

aednichols commented:
The design adapts pretty well to additional text, which is nice.

![screen shot 2017-12-04 at 1 12 50 pm](https://user-images.githubusercontent.com/1087943/33568608-081d0ae6-d8f5-11e7-9d69-ee682d67f350.png)
![screen shot 2017-12-04 at 1 13 21 pm](https://user-images.githubusercontent.com/1087943/33568609-08292812-d8f5-11e7-8271-8eb6a98a5527.png)

aednichols committed:
trialCurrentState -> trialState

To match backend changes in commit https://github.com/broadinstitute/firecloud-orchestration/pull/554/commits/848be4a3a682179ff8a1d7ddb00b6e2e2b55f547

davidangb commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/FreeTrialSpec.scala:
```diff
@@ -0,0 +1,76 @@
+      withSignIn(testUser) { _ =>
+        await ready new WorkspaceListPage()
+        val bannerTitle = Label(TestId("trial-banner-title")).getText
+        bannerTitle shouldBe ""
```
checking that the title is the empty string feels a little fragile - for instance, what if we are showing a banner but the underlying json in the bucket has an empty string? Could you test `Label(TestId("trial-banner-title")).isVisible` instead?

davidangb reviewed (commented):
some minor automation comments inline.

Functionally, I see a problem when the user is in `trialState:Disabled` ... I see a banner with no text. Disabled users should see no banner ... so the code can't just check for the presence of a `trialState` key.

Should probably add a test for the disabled state.

zarsky-broad committed:
choose user correctly

aednichols committed:
Disabled users do not see a banner

aednichols committed:
Merge branch 'iz-GAWB-2810-free-tier-banners' of github.com:broadinstitute/firecloud-ui into iz-GAWB-2810-free-tier-banners

zarsky-broad committed:
PR comments

zarsky-broad committed:
nit: spelling & spacing

zarsky-broad committed:
PR comments and more explicit testing

davidangb reviewed (approved):
:+1: code looks good to me, did not test.

zarsky-broad committed:
Select a user each time

zarsky-broad committed:
better labeling

zarsky-broad committed:
add TODO, don't require unneeded arguments

zarsky-broad commented:
jenkins retest

zarsky-broad committed:
better test naming, again
